contracts/utils/cryptography/ZKEmailUtils.sol

```diff
@@ -0,0 +1,96 @@
+ */
+library ZKEmailUtils {
+    using Strings for string;
+
+    /// @dev Enumeration of possible email proof validation errors.
+    enum EmailProofError {
+        NoError,
+        DKIMPublicKeyHash, // The DKIM public key hash verification fails
+        MaskedCommandLength, // The masked command length exceeds the maximum
+        SkippedCommandPrefixSize, // The skipped command prefix size is invalid
+        InvalidCommand, // The command format is invalid
+        EmailProof // The email proof verification fails
+    }
+
+    /**
+     * @dev Variant of {isValidZKEmail} that validates the `["signHash", "{uint}"]` command template.
+     */
+    function isValidZKEmail(
+        EmailAuthMsg memory emailAuthMsg,
+        IDKIMRegistry dkimregistry,
+        IVerifier verifier
+    ) internal view returns (EmailProofError) {
+        string[] memory signHashTemplate = new string[](2);
+        signHashTemplate[0] = "signHash";
+        signHashTemplate[1] = CommandUtils.UINT_MATCHER;
+
+        return isValidZKEmail(emailAuthMsg, dkimregistry, verifier);
+    }
+
+    /**
+     * @dev Validates a ZK-Email authentication message.
+     *
+     * Requirements:
+     *
+     * - The DKIM public key hash must be valid according to the registry
+     * - The masked command length must not exceed the verifier's limit
+     * - The skipped command prefix size must be valid
+     * - The command format and parameters must be valid
+     * - The email proof must be verified by the verifier
+     *
+     * This function takes an email authentication message, a DKIM registry contract, and a verifier contract
+     * as inputs. It performs several validation checks and returns a tuple containing a boolean success flag
+     * and an {EmailProofError} if validation failed. The function will return true with {EmailProofError.NoError}
+     * if all validations pass, or false with a specific {EmailProofError} indicating which validation check failed.
+     */
+    function isValidZKEmail(
+        EmailAuthMsg memory emailAuthMsg,
+        IDKIMRegistry dkimregistry,
+        IVerifier verifier,
+        string[] memory template
+    ) internal view returns (EmailProofError) {
+        if (!dkimregistry.isDKIMPublicKeyHashValid(emailAuthMsg.proof.domainName, emailAuthMsg.proof.publicKeyHash)) {
+            return EmailProofError.DKIMPublicKeyHash;
+        } else if (bytes(emailAuthMsg.proof.maskedCommand).length > verifier.commandBytes()) {
+            return EmailProofError.MaskedCommandLength;
+        } else if (emailAuthMsg.skippedCommandPrefix >= verifier.commandBytes()) {
+            return EmailProofError.SkippedCommandPrefixSize;
+        } else if (
+            !CommandUtils.computeExpectedCommand(emailAuthMsg.commandParams, template, 0).equal(
+                CommandUtils.removePrefix(emailAuthMsg.proof.maskedCommand, emailAuthMsg.skippedCommandPrefix)
```

Comment:
Note: `CommandUtils.removePrefix` is equivalent (but less efficient) to `Bytes.slice` for the openzeppelin/contracts

Reply:
`removePrefix` uses a for loop, `slice` does a `mcopy`.